src/js/abstract/hocs/with-base.js

```diff
@@ -0,0 +1,221 @@
+import camelize from '../../camelize';
+import debounce from '../../debounce';
+import getAttribute from '../../get-attribute';
+import maybe from '../../maybe';
+import PropertyExistsException from '../property-exists-exception';
+
+const THROWED_ERROR = 'throwed';
+const PROPERTY_WHITELIST = ['title', 'checked', 'disabled'];
+const ids = {};
+const getId = (nodeName) => {
+  if (!(nodeName in ids)) {
+    ids[nodeName] = 0;
+  }
+
+  return ++ids[nodeName]; // eslint-disable-line no-plusplus
+};
+const lifecycleLogger = maybe((...args) => console.log(...args))()(true);
+
+export const withBase = Base =>
+  class extends Base {
+    constructor(styles = '', template) {
+      super();
+
+      this._initialise(styles, template);
+      this._id = getId(this.nodeName);
+      this._props = {};
+      this.__attributeToPropertyMap = {};
+      this.reRender = debounce(() => this.render(), 50);
+
+      const { constructor: { observedAttributes } } = this;
+
+      // add DOM property getters/setters for related attributes
+      if (Array.isArray(observedAttributes)) {
+        observedAttributes.forEach((attr) => {
+          const key = camelize(attr);
+          this.__attributeToPropertyMap[attr] = key;
+          const hasKey = key in this;
+
+          if (ENV !== PROD) {
+            lifecycleLogger(this.logLifecycle)(`\n<-> apply getter/setter for ${key} by _${attr}`);
+          }
+
+          if (PROPERTY_WHITELIST.indexOf(key) === -1 && hasKey) {
+            throw new PropertyExistsException(key, this);
+          }
+
+          // @todo: may we should allow deletion by setting configurable: true
+          Object.defineProperty(this, key, {
+            get() {
+              return this[`_${attr}`];
+            },
+            set(value) {
+              this[`_${attr}`] = value;
+
+              this._props[key] = value;
+
+              if (hasKey) {
+                super[key] = value;
+              }
+
+              if (this._isConnected && this._hasRendered) {
+                if (ENV !== PROD) {
+                  lifecycleLogger(this.logLifecycle)(`\n---> setter for ${key} by _${attr}`);
+                }
+
+                this.reRender();
+              }
+            },
+          });
+        });
+      }
+    }
+
+    /**
+     * _initialise - description
+     *
+     * @param  {type} styles description
+     * @param  {type} template description
+     * @return {type}        description
+     */
+    _initialise(styles, template = null) {
+      this._styles = styles;
```

Comment:
@TheDadi 
@LucaMele 
The `_styles` prop would belong to `withStyles` HOC i would say